File: api/src/api/models/PlaidItem.scala
package api.models

import java.time.Instant
import java.util.UUID

/** PlaidItemId is the internal database ID for a Plaid item. It is not the actual Plaid-generated item ID.
  */
opaque type PlaidItemId = UUID

object PlaidItemId:
    def apply(uuid: UUID): PlaidItemId = uuid

    extension (itemId: PlaidItemId)
        def toUUID: UUID = itemId

enum PlaidItemStatus:
    case Good,
        Bad,
        Unknown

    override def toString(): String =
        this match
            case Good =>
                "good"
            case Bad =>
                "bad"
            case Unknown =>
                "unknown"

object PlaidItemStatus:
    def fromString(s: String): PlaidItemStatus =
        s match
            case "good" =>
                PlaidItemStatus.Good
            case "bad" =>
                PlaidItemStatus.Bad
            case "unknown" =>
                PlaidItemStatus.Unknown

case class PlaidItem(
    id: PlaidItemId,
    userId: UUID,
    plaidAccessToken: String,
    plaidItemId: String,
    plaidInstitutionId: String,
    status: PlaidItemStatus,
    transactionsCursor: Option[String],
    createdAt: Instant,
    lastSyncedAt: Option[Instant],
    lastSyncError: Option[String],
    lastSyncErrorAt: Option[Instant],
    retryCount: Int,
    errorType: Option[String],
    errorCode: Option[String],
    errorMessage: Option[String],
    errorDisplayMessage: Option[String],
    errorRequestId: Option[String],
    documentationUrl: Option[String],
    suggestedAction: Option[String]
)

File: finny-worker/plaid_item/plaid_item.go
package plaid_item

import (
	"errors"
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type PlaidItemRepository struct {
	db *gorm.DB
}

func NewPlaidItemRepository(db *gorm.DB) *PlaidItemRepository {
	return &PlaidItemRepository{
		db: db,
	}
}

func (r *PlaidItemRepository) GetItem(itemId uuid.UUID) (*PlaidItem, error) {
	var item PlaidItem
	err := r.db.Where("id = ?", itemId).First(&item).Error
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, nil
		}
		return nil, err
	}
	return &item, nil
}

type PlaidItem struct {
	ID                  uuid.UUID      `gorm:"type:uuid;default:gen_random_uuid();primaryKey"`
	UserID              uuid.UUID      `gorm:"type:uuid;not null"`
	PlaidAccessToken    string         `gorm:"type:text;not null"`
	PlaidItemID         string         `gorm:"column:plaid_item_id;type:text;not null"`
	PlaidInstitutionID  string         `gorm:"type:text;not null"`
	Status              string         `gorm:"type:text;not null"`
	TransactionsCursor  *string        `gorm:"type:text"`
	CreatedAt           time.Time      `gorm:"type:timestamp(6) with time zone;not null;default:CURRENT_TIMESTAMP"`
	UpdatedAt           time.Time      `gorm:"type:timestamp(6) with time zone;not null;default:CURRENT_TIMESTAMP"`
	DeletedAt           gorm.DeletedAt `gorm:"type:timestamp(6) with time zone"`
	LastSyncedAt        *time.Time     `gorm:"type:timestamp with time zone"`
	LastSyncError       *string        `gorm:"type:text"`
	LastSyncErrorAt     *time.Time     `gorm:"type:timestamp with time zone"`
	RetryCount          int            `gorm:"type:integer;default:0"`
	ErrorType           *string        `gorm:"type:text"`
	ErrorCode           *string        `gorm:"type:varchar"`
	ErrorMessage        *string        `gorm:"type:varchar"`
	ErrorDisplayMessage *string        `gorm:"type:varchar"`
	ErrorRequestID      *string        `gorm:"type:varchar"`
	DocumentationURL    *string        `gorm:"type:varchar"`
	SuggestedAction     *string        `gorm:"type:varchar"`
}

func (PlaidItem) TableName() string {
	return "plaid_items"
}
